File: Part2Questions/q4_NumberOfIslands.java
package Part2Questions;
import java.util.*;
/**
 * Approach:
 * I treat the grid as a graph where each cell is a node.
 * When I find a '1', I use DFS to explore and mark all connected land cells as visited.
 * Each time I start a new DFS, it means I found a new island.
 *
 * DSA used: Graph + DFS
 *
 * Time Complexity: O(m * n) → visit each cell once
 * Space Complexity: O(m * n) → recursion stack in worst case
 * time spent 20min
 */
public class q4_NumberOfIslands {

    static int numOfIslands(int[][] islandMatrix){
        if(islandMatrix.length==0){
            return 0;//return 0 island
        }
        int countIsland = 0;
        for(int i=0;i<islandMatrix.length;i++){
            for(int j=0;j<islandMatrix[0].length;j++){
                if(islandMatrix[i][j]==1){
                    countAddress(islandMatrix,i,j);
                    countIsland++;
                }
            }
        }
        return countIsland;
    }
    private static void countAddress(int[][] islandMatrix, int row,int col){
        if(row < 0 || col <0 || row >= islandMatrix.length || col >= islandMatrix[0].length || islandMatrix[row][col]==0){
            return;
        }
        islandMatrix[row][col] = 0;

        countAddress(islandMatrix,row-1,col);
        countAddress(islandMatrix,row+1,col);
        countAddress(islandMatrix,row,col-1);
        countAddress(islandMatrix,row,col+1);
    }

    public static void main(String[] args){
        int[][] test1 ={
            {1,0,1,1,1},
            {1,1,0,1,1},
            {0,1,0,0,0},
            {0,0,0,1,0},
            {0,0,0,0,0}

        };
        int[][] test2 ={
            {1,0,0},
            {0,0,0}

        };
        System.out.println(numOfIslands(test2));
    }
    
}

File: homework3/jesus_romero/q4_NumberOfIslands.py
# Technique: Graph Traversal (DFS) on a 2D Grid

def numIslands(grid): # Time Complexity: O(R * C), Space Complexity: O(R * C)
    # 1. Handle edge case for empty input
    if not grid:
        return 0
    
    rows, cols = len(grid), len(grid[0])
    visited = set()
    island_count = 0

    # 2. Define a Depth-First Search (DFS) to explore contiguous land
    def dfs(r, c):
        # 3. Check boundaries and verify if cell is unvisited land (1)
        if (r < 0 or r >= rows or 
            c < 0 or c >= cols or 
            grid[r][c] == 0 or 
            (r, c) in visited):
            return
        
        # 4. Mark current coordinate as visited
        visited.add((r, c))
        
        # 5. Explore all 4 cardinal directions (up, down, left, right)
        dfs(r + 1, c)
        dfs(r - 1, c)
        dfs(r, c + 1)
        dfs(r, c - 1)

    # 6. Iterate through every coordinate in the grid
    for r in range(rows):
        for c in range(cols):
            # 7. Start a new DFS traversal only when encountering unvisited land
            if grid[r][c] == 1 and (r, c) not in visited:
                island_count += 1
                dfs(r, c)

    return island_count

# Time Complexity: O(M * N) - each cell is visited at most once.
# Space Complexity: O(M * N) - for the visited set and recursion stack in the worst case.
# Time spent: 35 minutes

class Test:
    def run_tests(self):
        # 1. Test Case: Multiple islands
        grid1 = [
            [1, 0, 1, 1, 1],
            [1, 1, 0, 1, 1],
            [0, 1, 0, 0, 0],
            [0, 0, 0, 1, 0],
            [0, 0, 0, 0, 0]
        ]
        assert numIslands(grid1) == 3
        
        # 2. Test Case: Single island
        grid2 = [
            [1, 0, 0],
            [0, 0, 0]
        ]
        assert numIslands(grid2) == 1
        
        # 3. Test Case: All water
        grid3 = [
            [0, 0, 0],
            [0, 0, 0]
        ]
        assert numIslands(grid3) == 0
        
        # 4. Test Case: All land
        grid4 = [
            [1, 1],
            [1, 1]
        ]
        assert numIslands(grid4) == 1
        
        print("NumIslands tests passed")

if __name__ == "__main__":
    Test().run_tests()